pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -0,0 +1,238 @@
+		glog.Errorf("PersistentVolumeController could not update volume %q: %+v", newVolume.Name, err)
+	}
+}
+
+// deleteVolume is callback from framework.Controller watching PersistentVolume
+// events.
+func (ctrl *PersistentVolumeController) deleteVolume(obj interface{}) {
+	// Intentionally left blank - we do not react on deleted volumes
+}
+
+// addClaim is callback from framework.Controller watching PersistentVolumeClaim
+// events.
+func (ctrl *PersistentVolumeController) addClaim(obj interface{}) {
+	claim, ok := obj.(*api.PersistentVolumeClaim)
+	if !ok {
+		glog.Errorf("Expected PersistentVolumeClaim but handler received %+v", obj)
+		return
+	}
+	if err := ctrl.syncClaim(claim); err != nil {
+		glog.Errorf("PersistentVolumeController could not add claim %q: %+v", claimToClaimKey(claim), err)
+	}
+}
+
+// updateClaim is callback from framework.Controller watching PersistentVolumeClaim
+// events.
+func (ctrl *PersistentVolumeController) updateClaim(oldObj, newObj interface{}) {
+	newClaim, ok := newObj.(*api.PersistentVolumeClaim)
+	if !ok {
+		glog.Errorf("Expected PersistentVolumeClaim but handler received %+v", newObj)
+		return
+	}
+	if err := ctrl.syncClaim(newClaim); err != nil {
+		glog.Errorf("PersistentVolumeController could not update claim %q: %+v", claimToClaimKey(newClaim), err)
+	}
+}
+
+// deleteClaim is callback from framework.Controller watching PersistentVolumeClaim
+// events.
+func (ctrl *PersistentVolumeController) deleteClaim(obj interface{}) {
+	var volume *api.PersistentVolume
+	if pvc, ok := obj.(*api.PersistentVolumeClaim); ok {
+		if pvObj, exists, _ := ctrl.volumes.store.GetByKey(pvc.Spec.VolumeName); exists {
+			if pv, ok := pvObj.(*api.PersistentVolume); ok {
+				volume = pv
+			}
+		}
+	}
+	if unk, ok := obj.(cache.DeletedFinalStateUnknown); ok && unk.Obj != nil {
+		if pv, ok := unk.Obj.(*api.PersistentVolume); ok {
+			volume = pv
+		}
+	}
+
+	// sync the volume when its claim is deleted.  Explicitly sync'ing the
+	// volume here in response to claim deletion prevents the volume from
+	// waiting until the next sync period for its Release.
+	if volume != nil {
+		err := ctrl.syncVolume(volume)
+		if err != nil {
+			glog.Errorf("PersistentVolumeController could not update volume %q from deleteClaim handler: %+v", volume.Name, err)
+		}
+	}
+}
+
+func (ctrl *PersistentVolumeController) syncVolume(volume *api.PersistentVolume) error {
+	glog.V(4).Infof("synchronizing PersistentVolume[%s], current phase: %s", volume.Name, volume.Status.Phase)
+
+	return nil
+}
+
+func (ctrl *PersistentVolumeController) syncClaim(claim *api.PersistentVolumeClaim) error {
+	glog.V(4).Infof("synchronizing PersistentVolumeClaim[%s], current phase: %s", claim.Name, claim.Status.Phase)
+
+	return nil
+}
+
+// Run starts all of this controller's control loops
+func (ctrl *PersistentVolumeController) Run() {
+	glog.V(4).Infof("starting PersistentVolumeController")
+	if ctrl.stopChannels == nil {
```

Comment:
Is this expected to ever be called more than once?  Why are you defending against repeat calls?

Alternatively, could this be simplified to:

```
if ctrl.stopChannels == nil {
    ctrl.stopChannels = make(map[string]chan struct{})

	ctrl.stopChannels["volumes"] = make(chan struct{})
    go ctrl.volumeController.Run(ctrl.stopChannels["volumes"])

	ctrl.stopChannels["claims"] = make(chan struct{})
 	go ctrl.claimController.Run(ctrl.stopChannels["claims"])
}